File: library/src/main/java/org/incendo/interfaces/pane/OrderedPane.java
package org.incendo.interfaces.pane;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.incendo.interfaces.element.Element;

import java.util.List;

public abstract class OrderedPane extends Pane {

    private final List<Integer> ordering;

    public OrderedPane(final List<Integer> ordering) {
        this.ordering = ordering;
    }

    public final List<Integer> ordering() {
        return this.ordering;
    }

    @Override
    public final @Nullable Element get(final int row, final int column) {
        return super.get(this.orderedRow(row), column);
    }

    @Override
    public final void put(final int row, final int column, final Element value) {
        super.put(this.orderedRow(row), column, value);
    }

    @Override
    public final boolean containsKey(final int row, final int column) {
        return super.containsKey(this.orderedRow(row), column);
    }

    private int orderedRow(final int row) {
        return this.ordering.get(row);
    }
}

File: next/src/main/java/org/incendo/interfaces/next/pane/OrderedPane.kt
package org.incendo.interfaces.next.pane

import org.incendo.interfaces.next.element.Element

public abstract class OrderedPane(
    private val ordering: List<Int>
) : Pane() {

    override fun get(row: Int, column: Int): Element? {
        return super.get(orderedRow(row), column)
    }

    override fun set(row: Int, column: Int, value: Element) {
        return super.set(orderedRow(row), column, value)
    }

    override fun has(row: Int, column: Int): Boolean {
        return super.has(orderedRow(row), column)
    }

    private fun orderedRow(row: Int) = ordering[row]
}
